File: scripts/bump_version.py
import sys
import os
import re

def bump_version():
    if len(sys.argv) < 2:
        print("Usage: python scripts/bump_version.py <new_version>")
        sys.exit(1)
        
    raw_version = sys.argv[1].strip()
    # Normalize version: strip leading 'v'
    if raw_version.startswith('v'):
        raw_version = raw_version[1:]
        
    # Validate semantic version format (X.Y.Z)
    semver_pattern = re.compile(r'^\d+\.\d+\.\d+(?:-[a-zA-Z0-9.]+)?(?:\+[a-zA-Z0-9.]+)?$')
    if not semver_pattern.match(raw_version):
        print(f"Error: '{sys.argv[1]}' is not a valid Semantic Version (e.g. 1.2.0, 0.1.0-alpha.1)")
        sys.exit(1)
        
    script_dir = os.path.dirname(os.path.abspath(__file__))
    root_dir = os.path.dirname(script_dir)
    
    # 1. Update VERSION file
    version_file = os.path.join(root_dir, "VERSION")
    with open(version_file, "w", encoding="utf-8") as f:
        f.write(raw_version + "\n")
    print(f"Updated VERSION to: {raw_version}")
    
    # 2. Update backend/pyproject.toml
    backend_toml = os.path.join(root_dir, "backend", "pyproject.toml")
    if os.path.exists(backend_toml):
        with open(backend_toml, "r", encoding="utf-8") as f:
            content = f.read()
        
        # Look for version = "..." under [project]
        new_content, count = re.subn(
            r'(^version\s*=\s*")[^"]+(")',
            rf'\g<1>{raw_version}\g<2>',
            content,
            flags=re.MULTILINE
        )
        if count > 0:
            with open(backend_toml, "w", encoding="utf-8", newline="\n") as f:
                f.write(new_content)
            print(f"Updated backend/pyproject.toml to: {raw_version}")
        else:
            print("Warning: Could not find version key in backend/pyproject.toml")
            
    # 3. Update frontend/package.json
    frontend_json = os.path.join(root_dir, "frontend", "package.json")
    if os.path.exists(frontend_json):
        with open(frontend_json, "r", encoding="utf-8") as f:
            content = f.read()
            
        new_content, count = re.subn(
            r'("version"\s*:\s*")[^"]+(")',
            rf'\g<1>{raw_version}\g<2>',
            content
        )
        if count > 0:
            with open(frontend_json, "w", encoding="utf-8", newline="\n") as f:
                f.write(new_content)
            print(f"Updated frontend/package.json to: {raw_version}")
        else:
            print("Warning: Could not find version key in frontend/package.json")

if __name__ == "__main__":
    bump_version()

File: scripts/bump-version.mjs
import fs from "node:fs";
import path from "node:path";
import process from "node:process";
import { fileURLToPath } from "node:url";

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);
const rootDir = path.resolve(__dirname, "..");

function bumpVersion() {
  const args = process.argv.slice(2);
  if (args.length === 0) {
    console.error("Usage: npm run version:bump <new_version>");
    process.exit(1);
  }

  let rawVersion = args[0].trim();
  if (rawVersion.startsWith("v")) {
    rawVersion = rawVersion.slice(1);
  }

  const semverPattern = /^\d+\.\d+\.\d+(?:-[a-zA-Z0-9.]+)?(?:\+[a-zA-Z0-9.]+)?$/;
  if (!semverPattern.test(rawVersion)) {
    console.error(`Error: "${args[0]}" is not a valid Semantic Version (e.g. 0.4.0, 1.0.0-beta.1)`);
    process.exit(1);
  }

  // 1. Update VERSION file
  const versionFilePath = path.join(rootDir, "VERSION");
  fs.writeFileSync(versionFilePath, `${rawVersion}\n`, "utf-8");
  console.log(`✓ Updated VERSION to ${rawVersion}`);

  // 2. Update package.json
  const packageJsonPath = path.join(rootDir, "package.json");
  if (fs.existsSync(packageJsonPath)) {
    const pkg = JSON.parse(fs.readFileSync(packageJsonPath, "utf-8"));
    pkg.version = rawVersion;
    fs.writeFileSync(packageJsonPath, `${JSON.stringify(pkg, null, 2)}\n`, "utf-8");
    console.log(`✓ Updated package.json version to ${rawVersion}`);
  }

  // 3. Update src/components/Version.tsx
  const versionComponentPath = path.join(rootDir, "src", "components", "Version.tsx");
  if (fs.existsSync(versionComponentPath)) {
    let content = fs.readFileSync(versionComponentPath, "utf-8");
    content = content.replace(/(const VERSION = ")[^"]+(";)/, `$1${rawVersion}$2`);
    fs.writeFileSync(versionComponentPath, content, "utf-8");
    console.log(`✓ Updated src/components/Version.tsx to ${rawVersion}`);
  }

  // 4. Update src/components/__tests__/Version.test.tsx
  const versionTestPath = path.join(rootDir, "src", "components", "__tests__", "Version.test.tsx");
  if (fs.existsSync(versionTestPath)) {
    let testContent = fs.readFileSync(versionTestPath, "utf-8");
    testContent = testContent.replace(/(const VERSION = ")[^"]+(";)/, `$1${rawVersion}$2`);
    fs.writeFileSync(versionTestPath, testContent, "utf-8");
    console.log(`✓ Updated src/components/__tests__/Version.test.tsx to ${rawVersion}`);
  }

  console.log(`\nVersion successfully bumped to v${rawVersion}!`);
}

bumpVersion();
